Issue #15 Use status file to get the IP and remove to look for leasefile

praveenkumar opened the pull request:
We don't need to have now lookupIPFromLeasesFile since all the major
distro have libvirt > 2.0.0 and that was for the 1.2.6

Also we are adding 120sec wait time in 40 loop in 3 sec interval then
default one.

cfergeau commented on libvirt.go:
```diff
@@ -553,23 +540,45 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
+
+	defer conn.Close()
+
+	libVersion, err := conn.GetLibVersion()
```
We don't want to get/print libvirt version every time we call lookupIP(), it's not going to change over the lifetime of the application anyway.

cfergeau commented on libvirt.go:
```diff
@@ -553,23 +540,45 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
 
-	return ip, err
+	// TODO: for everything > 1002006, there is direct support in the libvirt-go for handling this
+	return d.getIPByMacFromSettings(mac)
```
libvirt-go is still parsing some internal libvirt files to achieve this. I believe the libvirt C library offers an API to directly get the IP rather than having to do it ourselves in libvirt.go

praveenkumar replied:
@cfergeau I don't have time right now to check in the C library implementation part and check if libvirt-go wrapper have same functionality. I can put a TODO here and we can circle back to it if that makes sense?

cfergeau commented on libvirt.go:
```diff
@@ -316,17 +316,28 @@ func (d *Driver) Start() error {
-	for i := 0; i < 90; i++ {
-		time.Sleep(time.Second)
-		ip, _ := d.GetIP()
+	for i := 0; i <= 40; i++ {
```
I would split all these changes from the getIPByMacFromSettings change as they are not related

praveenkumar replied:
Yes I will do that.

praveenkumar committed:
Use status file to get the IP and remove to look for leasefile

We don't need to have now lookupIPFromLeasesFile since all the major
distro have libvirt > 2.0.0 and that was for the 1.2.6

praveenkumar committed:
Get virtual-machine state before ask for IP

Right now we are directly asking for the IP instead
checking the VM state which can lead to false result.

praveenkumar committed:
Handle error for GetIP instead ignoring it

Once we create the VM  we are waiting for 5 seconds to atleast libvirt to kick in
and starting the VM, Also `if s != state.Running`  is mostly to catch other state
of the VM like crashed.

If that 5 sec is not enough for inital libvirt kick in we need to figure out what
could be better.

- https://github.com/code-ready/machine/blob/master/libmachine/state/state.go

praveenkumar committed:
Increase get IP retry time from 90s to 120s

Some of the platform the VM take more time to start and with 90s
limit it fails but the VM actually just need bit more time to start.
Increase in the IP retry will make VM start on bit slow systems or nested environments.

Add some logging message and corrected the conditions around ip
variable.

praveenkumar commented:
@cfergeau I broken the commit messages, have a look again.

cfergeau commented on libvirt.go:
```diff
@@ -558,15 +534,7 @@ func (d *Driver) GetIP() (string, error) {
-	}
-
+	ip, err := d.getIPByMacFromSettings(mac)
 	return ip, err
```
these 2 lines can be condensed in `return d.getIPByMacFromSettings(mac)`

cfergeau commented on libvirt.go:
```diff
@@ -530,6 +530,13 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
+	}
+	if s != state.Running {
+		return "", errors.New("host is not running")
+	}
```
what kind of false results do we get if the VM is not running?

praveenkumar replied:
It is not about false result but exit as soon as possible we don't even have to go to next step if the VM is not running.

cfergeau commented on libvirt.go:
```diff
@@ -318,7 +318,10 @@ func (d *Driver) Start() error {
+			return fmt.Errorf("%v: getting ip during machine start", err)
+		}
 		if ip != "" {
 			// Add a second to let things settle
```
I think this loop is ignoring errors on purpose, and is busy looping while it waits for the VM to start. Do you have a specific error in mind when we could return early because we know we are never recover from that error?

cfergeau replied:
Still the same comment, I believe the whole point of the loop is to wait until everything is up. If you are reasonably sure that we should be looping while we get `"", nil` from `GetIP()`, but that if we get `"", err` from `GetIP()`, then this is never going to be a transient error (ie the error won't go away once the boot process has progressed enough), then please explain this in the commit log.

cfergeau replied:
No input on this one?

cfergeau commented on libvirt.go:
```diff
@@ -316,20 +316,29 @@ func (d *Driver) Start() error {
 	// They wont start immediately
```
Not clear from the commit log why it's desirable to retry for 120s rather than 90s?

cfergeau commented on libvirt.go:
```diff
@@ -316,20 +316,29 @@ func (d *Driver) Start() error {
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
```
I would move this block to the very end of the loop where we currently have `log.Debugf("Waiting for the VM to come up... %d", i)` this should make the code easier to follow

praveenkumar replied:
@cfergeau this block is already correct place.

```golang
if ip == "" {
			log.Debugf("Waiting for machine to come up %d/%d", i, 40)
			time.Sleep(3 * time.Second)
			continue
 }
```

cfergeau replied:
if you move it to the end of the loop, you can get rid of the `if ip == ""` and of the `continue` without (I think) any changes of when this code is run. This makes the code easier to follow.

praveenkumar replied:
@cfergeau Now I got it what you meant, will update it.

cfergeau commented on libvirt.go:
```diff
@@ -316,20 +316,29 @@ func (d *Driver) Start() error {
-			time.Sleep(time.Second)
-			return nil
+			log.Infof("Found IP for machine: %s", ip)
+			d.IPAddress = ip
```
ah ok this also adds some caching of the IP address in the Driver structure. Are you sure we don't need to keep the 'sleep' though?

praveenkumar replied:
@cfergeau once we get the IP we don't need any sleep and return it to driver.

cfergeau replied:
I don't know, there was a comment implying it's desirable to wait a little bit more after getting the IP. I'm worried by not waiting anymore, we'll be introducing some races ?

praveenkumar replied:
Alright if that make you bit uncomfortable, I will add that sleep :)

cfergeau commented on libvirt.go:
```diff
@@ -316,20 +316,28 @@ func (d *Driver) Start() error {
 		}
-		log.Debugf("Waiting for the VM to come up... %d", i)
+
+		log.Debugf("Waiting for machine to come up %d/%d", i, 40)
```
This is starting to be very small details, but it would be better to define a const for the number of loops we are going to run, as if we want to change the number of iterations, we are bound to forget this one ;)

cfergeau commented on libvirt.go:
```diff
@@ -316,20 +316,28 @@ func (d *Driver) Start() error {
+
+		log.Debugf("Waiting for machine to come up %d/%d", i, 40)
+		time.Sleep(3 * time.Second)
+		continue
```
The continue can be removed.

cfergeau reviewed (approved):
I still would like more details in the commit log for `Handle error for GetIP instead ignoring it` (what we discussed, why this is not going to introduce a regression). Apart from this, it looks good to me.
